File: 23.go
package main

import (
	"bufio"
	"fmt"
	"os"
)

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

type pos struct{ x, y int }

func main() {
	scanner := bufio.NewScanner(os.Stdin)
	elves := map[pos]bool{}
	for y := 0; scanner.Scan(); y++ {
		for x, c := range scanner.Text() {
			if c == '#' {
				elves[pos{x, y}] = true
			}
		}
	}
	n1, n2 := 0, 0
	for i := 0; true; i++ {
		propositions := map[pos]pos{}
		for e := range elves {
			step := false
			for _, d := range []pos{{0, -1}, {0, 1}, {-1, 0}, {1, 0}, {-1, -1}, {1, 1}, {-1, 1}, {1, -1}} {
				if elves[pos{e.x + d.x, e.y + d.y}] {
					step = true
					break
				}
			}
			if step {
				moves := []struct {
					to   pos
					scan []pos
				}{
					{pos{e.x, e.y - 1}, []pos{{e.x, e.y - 1}, {e.x - 1, e.y - 1}, {e.x + 1, e.y - 1}}},
					{pos{e.x, e.y + 1}, []pos{{e.x, e.y + 1}, {e.x - 1, e.y + 1}, {e.x + 1, e.y + 1}}},
					{pos{e.x - 1, e.y}, []pos{{e.x - 1, e.y}, {e.x - 1, e.y - 1}, {e.x - 1, e.y + 1}}},
					{pos{e.x + 1, e.y}, []pos{{e.x + 1, e.y}, {e.x + 1, e.y - 1}, {e.x + 1, e.y + 1}}},
				}
				for j := 0; j < 4; j++ {
					to := (i + j) % 4
					scan := moves[to].scan
					if !elves[scan[0]] && !elves[scan[1]] && !elves[scan[2]] {
						propositions[e] = moves[to].to
						break
					}
				}
			}
		}
		nproposition := map[pos]int{}
		for _, to := range propositions {
			nproposition[to]++
		}
		moved := map[pos]bool{}
		for p, to := range propositions {
			if nproposition[to] == 1 {
				moved[to] = true
				delete(elves, p)
			}
		}
		if len(moved) == 0 {
			n2 = i + 1
			break
		}
		for e := range elves {
			moved[e] = true
		}
		elves = moved
		if i == 9 {
			top, left, bottom, right := 1000000, 1000000, -1000000, -1000000
			for e := range elves {
				top, left, bottom, right = min(top, e.x), min(left, e.y), max(bottom, e.x), max(right, e.y)
			}
			n1 = (bottom-top+1)*(right-left+1) - len(elves)
		}
	}
	fmt.Println(n1, n2)
}

File: 23.swift
import Foundation

struct Pos: Hashable {
    var x, y: Int

    init(_ x: Int, _ y: Int) { self.x = x; self.y = y; }
}

var elves = Set<Pos>()
var y = 0
while let line = readLine() {
    for (x, c) in line.enumerated() {
        if c == "#" { elves.insert(Pos(x, y)) }
    }
    y += 1
}
var (n1, n2, i) = (0, 0, 0)
while true {
    var propositions = [Pos: Pos]()
    for e in elves {
        if elves.contains(where :{ $0 == Pos(e.x, e.y - 1) || $0 == Pos(e.x, e.y + 1) ||
                            $0 == Pos(e.x - 1, e.y) || $0 == Pos(e.x + 1, e.y) ||
                            $0 == Pos(e.x - 1, e.y - 1) || $0 == Pos(e.x + 1, e.y + 1) ||
                            $0 == Pos(e.x - 1, e.y + 1) || $0 == Pos(e.x + 1, e.y - 1) }) {
            let moves = [
                [Pos(e.x, e.y - 1), Pos(e.x, e.y - 1), Pos(e.x - 1, e.y - 1), Pos(e.x + 1, e.y - 1)],
    			[Pos(e.x, e.y + 1), Pos(e.x, e.y + 1), Pos(e.x - 1, e.y + 1), Pos(e.x + 1, e.y + 1)],
                [Pos(e.x - 1, e.y), Pos(e.x - 1, e.y), Pos(e.x - 1, e.y - 1), Pos(e.x - 1, e.y + 1)],
                [Pos(e.x + 1, e.y), Pos(e.x + 1, e.y), Pos(e.x + 1, e.y - 1), Pos(e.x + 1, e.y + 1)]
            ]
            for j in 0..<4 {
                let to = (i + j) % 4
                if !elves.contains(moves[to][1]) && !elves.contains(moves[to][2]) && !elves.contains(moves[to][3]) {
                    propositions[e] = moves[to][0]
                    break
                }
            }
        }
    }
    var npropositions = [Pos:Int]()
    for (from, to) in propositions {
        npropositions[to] = (npropositions[to] ?? 0) + 1
    }
    var moved = 0
    for (from, to) in propositions {
        if npropositions[to] == 1 {
            elves.remove(from)
            elves.insert(to)
            moved += 1
        }
    }
    if moved == 0 {
        n2 = i + 1
        break
    }
    if i == 9 {
        var top = 1000000, left = 1000000, bottom = -1000000, right = -1000000
        for e in elves {
            (left, right, top, bottom) = (min(left, e.x), max(right, e.x), min(top, e.y), max(bottom, e.y))
        }
        n1 = (bottom - top + 1) * (right - left + 1) - elves.count
    }
    i += 1
}
print(n1, n2)
